### ISSUE TITLE
Browser error that ASP.NET Core dev certificate is not trusted if browser was already open when trust was established

### ISSUE THREAD

**Issue by danroth27** [dotnet/aspnetcore#49606]
Users are complaining that they are getting browser errors complaining about the ASP.NET dev certificate even after it was trusted. It appears that something changed with browsers that they don't pick up the certificate trust if the browser was open when trust was established.

**Comment by jinzhao1127** [comment]
This issue similar as https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1801332

**Comment by ericlaw1979** [comment]
Cert handling /did/ change in Chromium which could conceivably explain, but given the lack of howling, it seems unlikely.
https://textslashplain.com/2022/12/06/tls-certificate-verification-changes-in-edge/

In terms of what to do, visit about://net-export and reproduce the problem. Then, look at the log file to see what the exact certificate handling looks like. If there really is an issue, https://crbug.com/new is the place to share the log to have Chromium folks look at it.

**Comment by danroth27** [comment]
@ericlaw1979 The issue only appears to happen if the certificate is added to the trusted root store while the browser is still open. If you close all open browser instances and then launch the site in a new browser instance, then the user doesn't see a certificate error. As far as I know, this wasn't previously the case. Unfortunately, new .NET users often have a browser instance open when they're trying .NET for the first time, so they run into this error and then think the app is somehow broken.

**Comment by danroth27** [comment]
> In terms of what to do, visit about://net-export and reproduce the problem. Then, look at the log file to see what the exact certificate handling looks like

@ericlaw1979 I did this and captured a log but scanning through JSON log file I'm not really sure what I'm looking for.

**Comment by ericlaw1979** [comment]
Investigation of the log reveals that Chromium is trying to build the certificate chain **twice**, once with the self-signed `localhost` certificate properly chained to itself, and once with that certificate chained to an _unrelated_ `localhost` certificate, which fails because the second certificate did not sign the first.

![image](https://github.com/dotnet/aspnetcore/assets/4192774/6a084ad9-d131-40d2-9516-9b72a84d531b)

I'm wondering whether the presence of that second `localhost` cert is what's causing the fallback logic to fail. 

Using `CertMgr.msc`, try investigating the trust stores to see if you have more than one `localhost` cert.

If you're generating the certificate yourself, it might also work to generate a certificate that contains an `Authority Key Identifier` field such that the chainbuilder does not get confused and think the server's leaf certificate is not self-signed.

**Comment by DamianEdwards** [comment]
>Using CertMgr.msc, try investigating the trust stores to see if you have more than one localhost cert.

Just FYI our dev cert is installed in the user store so isn't shown in `CertMgr.msc` but rather one has to manually add the user store to a new MMC instance (IIUC).

**Comment by DamianEdwards** [comment]
On Windows machines we do usually end up with two `localhost` self-signed certs. One for IIS Express, and another for ASP.NET Core. When launching via ASP.NET Core's server (Kestrel) directly (which is the default), obviously the IIS Express cert will not be relevant.

**Comment by DamianEdwards** [comment]
I tried reproducing this issue and not having the second `localhost` certificate didn't seem to make any difference. In all variations I tried, only restarting the browser resulted in the certificate being trusted.

**Comment by amcasey** [comment]
The VS feedback ticket says it happens in Edge but not Chrome.  Do we think that's just because they probably started a fresh chrome but already had an edge instance open?

While I agree with the argument that we'd expect more screaming for a general browser issue, I'm not sure I understand how this could be specific to our cert, since our cert doesn't (AFAIK) change between the run that doesn't work and the run that does.  Unless there's some metadata we "recently" started including/omitting that forces a restart before the cert works (unlikely?).

**Comment by amcasey** [comment]
> I'm wondering whether the presence of that second localhost cert is what's causing the fallback logic to fail.

That's a compelling theory, but why would that problem go away the next time chrome is started?  Maybe one has already been cached and the other succeeds on its own or it simply doesn't matter that the second failed?

**Comment by ericlaw1979** [comment]
I believe the root cause of this is well understood at this point.

Upstream this is tracked by https://crbug.com/1475411 ; internally Celeste Pan on Edge Networking has the issue in her queue.

**Comment by amcasey** [comment]
Thanks for the quick response!

Those are some heavy repro steps. 😆 You could probably get away with just the dotnet SDK and not all of VS.

> would it make sense to introduce a single retry inside Chromium when running on Windows?

😢

**Comment by amcasey** [comment]
@danroth27 Any objection to closing this now that it's tracked externally?  Sounds like it should start working for people as soon as the fix is available.

**Comment by DamianEdwards** [comment]
I think we should leave this issue open to track the external work given the initial report was via ASP.NET Core users.

**Comment by amcasey** [comment]
> I think we should leave this issue open to track the external work given the initial report was via ASP.NET Core users.

Makes sense.  In that case, I'll edit the description to link to the external issue.

**Comment by ericlaw1979** [comment]
> Those are some heavy repro steps. 😆 You could probably get away with just the dotnet SDK and not all of VS.

Yes, this is an unfortunate habit when Microsoft folks file bugs in Chromium. I didn't object in this particular case because anything Windows-specific is already a burden for Chrome engineers (most of them don't run Windows) and we knew that Edge engineers were likely to be doing the bulk of the investigation and likely making the fix anyway.

**Comment by vip32** [comment]
just a heads up from 2025, just ran into this issue (untrusted cert + renewal) and can confirm that it is still the case with Edge currently (Version 140.0.3485.81)

**Comment by ericlaw1979** [comment]
@vip32: the fix landed recently in 141:
https://chromiumdash.appspot.com/commit/bc1d2b80c5380a114d343223fcdcd3cbd55de701
